Bottom: 54ed7a1b33b21909cebb0f98dfa5bdd3674e03a2
Top:    c1ee2a0fc4e3fd3ddbfb973050604b30042a9aae
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2016-12-04 19:20:03 +0000

Fixup code to determine whether we draw pins/vias on a given layer

The net change of this patch is that we consider whether to draw pins/vias
for each layer within the layer group. Previously, if we encoundered a layer
named "outline" or "route" within the layer group, all _subsequent_ layers
within that group would not have their pins/vias rendered.

As ordering of the layers within the group should not be important like this,
change behaviour to apply test on a per-layer basis.


---

diff --git a/src/hid/gtk/gtkhid-gl.c b/src/hid/gtk/gtkhid-gl.c
index 02de97b..0f605e1 100644
--- a/src/hid/gtk/gtkhid-gl.c
+++ b/src/hid/gtk/gtkhid-gl.c
@@ -1408,7 +1408,7 @@ static int
 GhidDrawLayerGroup (int group, const BoxType * screen)
 {
   render_priv *priv = gport->render_priv;
-  int i, rv = 1;
+  int i;
   int layernum;
   int side;
   struct poly_info info;
@@ -1418,6 +1418,7 @@ GhidDrawLayerGroup (int group, const BoxType * screen)
   int first_run = 1;
   int top_group = GetLayerGroupNumberBySide (TOP_SIDE);
   int bottom_group = GetLayerGroupNumberBySide (BOTTOM_SIDE);
+  bool is_outline;
 
   if (!hid_draw_set_layer (&ghid_graphics, 0, group, 0))
     return 0;
@@ -1427,9 +1428,8 @@ GhidDrawLayerGroup (int group, const BoxType * screen)
     layernum = layers[i];
     Layer = PCB->Data->Layer + layers[i];
 
-    if (strcmp (Layer->Name, "outline") == 0 ||
-        strcmp (Layer->Name, "route") == 0)
-      rv = 0;
+    is_outline = strcmp (Layer->Name, "outline") == 0 ||
+                 strcmp (Layer->Name, "route") == 0;
 
     if (layernum < max_copper_layer && Layer->On) {
 
@@ -1438,7 +1438,7 @@ GhidDrawLayerGroup (int group, const BoxType * screen)
 
       first_run = 0;
 
-      if (rv && !TEST_FLAG (THINDRAWFLAG, PCB)) {
+      if (!is_outline && !TEST_FLAG (THINDRAWFLAG, PCB)) {
         /* Mask out drilled holes on this layer */
         hidgl_flush_triangles (priv->hidgl);
         glPushAttrib (GL_COLOR_BUFFER_BIT);
@@ -1461,7 +1461,7 @@ GhidDrawLayerGroup (int group, const BoxType * screen)
         hid_draw_end_layer (&ghid_graphics);
         hid_draw_set_layer (&ghid_graphics, 0, group, 0);
 
-        if (rv && !TEST_FLAG (THINDRAWFLAG, PCB)) {
+        if (!is_outline && !TEST_FLAG (THINDRAWFLAG, PCB)) {
           hidgl_flush_triangles (priv->hidgl);
           glPushAttrib (GL_COLOR_BUFFER_BIT);
           glColorMask (0, 0, 0, 0);
@@ -1474,7 +1474,7 @@ GhidDrawLayerGroup (int group, const BoxType * screen)
       }
 
       /* Draw pins, vias and pads on this layer */
-      if (!global_view_2d && rv) {
+      if (!global_view_2d && !is_outline) {
         if (PCB->PinOn) r_search (PCB->Data->pin_tree, screen, NULL, pin_inlayer_callback, Layer);
         if (PCB->ViaOn) r_search (PCB->Data->via_tree, screen, NULL, via_inlayer_callback, Layer);
         if (PCB->PinOn && group == top_group)
